### ISSUE TITLE
Memory Leaking with pdf viewer

### ISSUE THREAD

**Issue by JoshuaDeanEssium** [mozilla/pdf.js#10021]
Hi, so i am using bower, requireJS, backboneJS, and i believe thats all thats necessary to know in my web software. I am trying to integrate this viewer and am including everything necessary (pdf.js, worker, and web/viewer). 

The issue is that no matter what i delete and remove the viewer still holds memory. Is there a way to tell the pdf_viewer to remove all references to children so that they are removed from memory?
![heap](https://user-images.githubusercontent.com/34247427/44748220-39dbca80-aadd-11e8-877b-1d2d6b3b8fde.PNG)
![destroy](https://user-images.githubusercontent.com/34247427/44748221-39dbca80-aadd-11e8-8001-8b1589792b91.PNG)
![howimloading](https://user-images.githubusercontent.com/34247427/44748222-39dbca80-aadd-11e8-8fd3-eeaf110985d3.PNG)

I create a new viewer when the view containing the pdf display is created (which is necessary for our software). So i need it to clear itself out completely on destroy since everything is ajax and asynchronous.

 I have gotten it down to a small amount of memory leaking but it is still holding onto old references. Attached are heap screenshots and code from what i am doing. Please let me know what else i need to call to remove the appropriate memory. 

Thank you

**Comment by JoshuaDeanEssium** [comment]
This is after going out of the view that generates the viewer 4 times (so it created 4 PDFViewer instances). I delete them and I even go through the pages and delete the annotationLayerFactory, TextLayerFactory, and Rendering queue. As well as the linkService and renderingQueue of the viewer. As you can see it is still holding memory and does not free anything (even when "deleting" them). Please help.
![comparison](https://user-images.githubusercontent.com/34247427/44800003-57636f80-ab83-11e8-8a29-e22768e882f7.PNG)

**Comment by JoshuaDeanEssium** [comment]
This is the bower file in the pdfjs-dist that i just integrated last week
"name": "pdfjs-dist",
  "version": "2.0.550",
  "main": [
    "build/pdf.js",
    "build/pdf.worker.js"
  ],
  "ignore": [],
  "keywords": [
    "Mozilla",
    "pdf",
    "pdf.js"
  ]

**Comment by prasadnagabathula** [comment]
Hi Tim,
I'm using pdf.js 1.4.20 and opening the PDF file in a modal window in an asp.net application. At production environment, whenever there are multiple documents opened simultaneously by the users, stackoverflow exceptions are occurring continuously.  The following is the error:

Faulting application name: w3wp.exe, version: 8.5.9600.16384, time stamp: 0x52157ba0
Faulting module name: KERNEL32.DLL, version: 6.3.9600.17415, time stamp: 0x545049be
Exception code: **0xc00000fd**
Fault offset: 0x0001bb0b
Faulting process id: 0x14bc
Faulting application start time: 0x01d453e495a2c365
Faulting application path: C:\Windows\SysWOW64\inetsrv\w3wp.exe
Faulting module path: C:\Windows\SYSTEM32\KERNEL32.DLL
Report Id: d3d7358f-bfd7-11e8-80e1-00505695600f
Faulting package full name: 
Faulting package-relative application ID: 

Are there any issues with memory usage if multiple documents are opened and accessed and even after closing the documents, the reference are not released or heap is not cleared. If yes, is it addressed in new versions of PDF.js 1.9.x or so?

Please help me to identify and resolve the issue.

Thank you..

**Comment by stephanrauh** [comment]
As far as I can see, there a quite a few memory leaks in the viewer. After removing the PDF viewer from the DOM and tidying up, the memory snapshot looks like this:

![image](https://user-images.githubusercontent.com/3045767/49438976-2026df00-f7c0-11e8-87f7-f09d4193435c.png)

**Comment by sgudivaka** [comment]
any update on this?

**Comment by stephanrauh** [comment]
Unfortunately not. It's a difficult topic, and I've got almost no free time this month. I hope to be able to return to the project soon.

You don't happen to have both free time and a lot of curiosity? I always appreciate pull requests! Or hints, to put the bar a bit lower :).

**Comment by JoshuaDeanEssium** [comment]
I have gotten memory to drop low enough that it isnt a huge problem but it still leaks a little memory so it would be nice to have a fix to this sometime.

**Comment by jboutin** [comment]
Possible dupe of #9902 

Both seem to be related to a memory leak...

**Comment by brendandahl** [comment]
If someone has a simple example that I can run locally that demonstrates this leak, it will be much easier to look into.

**Comment by JoshuaDeanEssium** [comment]
Unfortunately i do not. If you create with PDFJS.getDocument then cleanup and remove from DOM then run it again with PDFJS.getDocument you will see leaks in dev tools -> memory.

I provided all the code necessary to ensure that it leaks in the initial post. 

To those that want my temporary solution for less memory leaks; I create the viewer only once then reuse that same viewer throughout the software so the only thing that leaks is a little PDFDocument memory each time it loads a new document. Since PDFViewer was the main culprit of the leaking I dont remove from memory once its created and just use the pointer to that viewer everywhere we display PDFs.

**Comment by dschissler** [comment]
@JoshuaDeanEssium How do you gain access to the worker `this.worker`?

**Comment by dschissler** [comment]
I'm using `2.1.266` (pre-release) and boy-oh-boy does it leak.  This is an app destroying blocker for me.  The good news for Mozilla is that Firefox 66 seems to be able to handle the app memory blowing up like a champ as well as being apparently able to release more/some memory than Chromium 74.

To duplicate this you simply load multiple PDFs, while clearing them.

I suspect that many of the serious pdf.js developers are using Firefox most of the time and so haven't noticed the memory leak / high memory as a blocker because Firefox just handles it better.  Quite simply if you want to catch this then use another browser because its so horrendous that this couldn't become anything other than a top priority.

**Comment by MichaelZelensky** [comment]
I am using version 2.2.15, and it still suffers from this issue. It renders the whole library simply unusable for documents larger than a number of pages, specific to a client machine. In my case (Win10/Chrome) the limit it is ~200 pages before it crashes the browser.

**Comment by timvandermeij** [comment]
Labeling as a performance issue for the viewer so we can find it more easily.

Perhaps you could try the latest development version, i.e., version 2.2.189 as found on https://mozilla.github.io/pdf.js/web/viewer.html or the latest `master` checkout, to see if it helps. Some fixes were done very recently with regards to less object creation (intermediate strings and primitive objects) and cache clearing (images and primitive objects), so perhaps that improves the situation a bit for the "PDFObjects"

In the memory usage report I do see a lot of page proxy and page view objects. From the screenshot in https://github.com/mozilla/pdf.js/issues/10021#issuecomment-417007360 I can at least say that the `PDFObjects` is kept alive by the `PDFPageProxy` and the `PDFPageProxy` is most likely kept alive in https://github.com/mozilla/pdf.js/blob/fef86cc3e327f0554136482fee901b52909fa37a/src/display/api.js#L2213. Perhaps this is a start for looking into where the cleanup is going wrong.

**Comment by mamichels** [comment]
Can report the same/similar issue (using Firefox 68.0.1 32-Bit) when embedding pdf files into the html through the object/embedded tag.
I am subsequently loading a lot of different pdf files(~200kb) while staying on the same site, while only one PDF is rendered at any given time. I am not sure why the garbage collector is neglecting them in the first place since the isn't any reference to the file after loading another one. Even the IE11 is handling the situation as expected.

![task](https://user-images.githubusercontent.com/48212402/62547924-542e3480-b866-11e9-9eb9-57eebcb4f257.JPG)
![memory](https://user-images.githubusercontent.com/48212402/62547926-555f6180-b866-11e9-85d0-f1ba365cbc43.JPG)

**Comment by ANSOLINC** [comment]
Having Same Problem in Firefox.  Is the Promise or wrapper preventing garbage collection?

![FireFox_PDFJS_Memory](https://user-images.githubusercontent.com/6362949/66453992-49de1180-ea34-11e9-9a59-7c25ff557673.png)

**Comment by moonrailgun** [comment]
Hi, is any update for this issue?

**Comment by dschissler** [comment]
I'm not sure, but all of my problems went away once I stopped looking for problems.  I mean that logging to the console was creating memory leaks.  So try deleting references in the code that you don't need and also then check without logging any pdf.js objects.

**Comment by Snuffleupagus** [comment]
Looking through this issue, there's *unfortunately* still a lack of any really actionable information here!

So far, only screen-shots have been provided, and it's unfortunately *impossible* to do any meaningful debugging without *runnable* code; please note https://github.com/mozilla/pdf.js/issues/10021#issuecomment-452920974 which explicitly asked for a test-case (more than two years ago) and also https://github.com/mozilla/pdf.js/blob/master/.github/CONTRIBUTING.md (emphasis mine):
> If you are developing a custom solution, first check the examples at https://github.com/mozilla/pdf.js#learning and search existing issues. If this does not help, **please prepare a short well-documented example that demonstrates the problem and make it accessible online** on your website, JS Bin, GitHub, etc. before opening a new issue or contacting us in the Matrix room -- **keep in mind that just code snippets won't help us troubleshoot the problem**.

---

Given the current information, or rather lack thereof, I'd suggest closing this issue as INCOMPLETE since it's simply impossible to do anything about this issue (or any other one for that matter) without a way to reproduce it.

(Finally, please note that we're *not* saying that there can't be memory leak somewhere in the PDF.js code-base, it's just that it's impossible to tell without a lot more information.)
